File: extension/src/lib/logger.ts
/**
 * Centralized logging utility for Roam extension.
 *
 * Prevents sensitive data leakage to:
 * 1. Browser console (user can inspect)
 * 2. Service Worker logs (persists indefinitely)
 * 3. Extension background page storage (accessible via chrome://extensions/details/)
 *
 * All errors automatically sent to Sentry for production monitoring.
 *
 * Usage in background.ts or popup.ts:
 *   import { logInfo, logError } from '@/lib/logger';
 *   logInfo('auth', 'OAuth initiated');
 *   logError('roam', 'Failed to fetch URL', { statusCode: 404 });
 */

import * as Sentry from '@sentry/browser';

export enum LogLevel {
  DEBUG = 0,
  INFO = 1,
  WARN = 2,
  ERROR = 3,
}

/**
 * Determine the effective log level.
 * Extension build mode: use BUILD_MODE env var
 * - development: DEBUG
 * - production: ERROR
 */
function getLogLevel(): LogLevel {
  // Check chrome.storage for persisted log level override
  let level = LogLevel.ERROR; // default to production-safe level

  if (typeof chrome !== 'undefined' && chrome.storage) {
    // Sync check (not ideal, but necessary during init)
    try {
      // Note: chrome.storage.local.get is async, so we can't use it here synchronously
      // Instead, init will call setLogLevelIfDev() to override this if needed
    } catch (e) {
      // Ignore errors
    }
  }

  // In development, default to INFO level for better debugging
  if (process.env.NODE_ENV === 'development') {
    level = LogLevel.INFO;
  }

  return level;
}

let CURRENT_LOG_LEVEL = getLogLevel();

/**
 * Safely format context, stripping sensitive fields.
 */
function sanitizeContext(context?: Record<string, any>): Record<string, any> {
  if (!context) return {};

  const safe: Record<string, any> = {};
  const SAFE_KEYS = ['statusCode', 'count', 'duration', 'retry', 'attempt', 'reason', 'category', 'type', 'action', 'platform'];
  const UNSAFE_KEYS = ['email', 'password', 'token', 'secret', 'userId', 'id', 'url', 'response', 'body', 'payload', 'session'];

  for (const [key, value] of Object.entries(context)) {
    if (UNSAFE_KEYS.some((unsafe) => key.toLowerCase().includes(unsafe))) continue;
    if (SAFE_KEYS.some((safe) => key.toLowerCase().includes(safe))) {
      safe[key] = value;
    }
  }

  return safe;
}

/**
 * Debug level log (only in development).
 */
export function logDebug(module: string, message: string, context?: Record<string, any>): void {
  if (CURRENT_LOG_LEVEL > LogLevel.DEBUG) return;

  const sanitized = sanitizeContext(context);
  console.debug(`[${module}] ${message}`, sanitized);
}

/**
 * Info level log.
 */
export function logInfo(module: string, message: string, context?: Record<string, any>): void {
  if (CURRENT_LOG_LEVEL > LogLevel.INFO) return;

  const sanitized = sanitizeContext(context);
  console.info(`[${module}] ${message}`, sanitized);
  Sentry.captureMessage(`[INFO] [${module}] ${message}`, 'info');
}

/**
 * Warning level log.
 */
export function logWarn(module: string, message: string, context?: Record<string, any>): void {
  if (CURRENT_LOG_LEVEL > LogLevel.WARN) return;

  const sanitized = sanitizeContext(context);
  console.warn(`[${module}] ${message}`, sanitized);
  Sentry.captureMessage(`[WARN] [${module}] ${message}`, 'warning');
}

/**
 * Error level log (always captured to Sentry).
 */
export function logError(module: string, message: string, context?: Record<string, any>, error?: Error): void {
  const sanitized = sanitizeContext(context);
  console.error(`[${module}] ${message}`, sanitized, error);

  if (error) {
    Sentry.captureException(error, {
      tags: { module, message },
      extra: sanitized,
    });
  } else {
    Sentry.captureMessage(`[ERROR] [${module}] ${message}`, 'error');
  }
}

/**
 * Override log level at runtime (for debugging).
 * Called from background.ts during init if in development mode.
 */
export function setLogLevel(level: LogLevel): void {
  CURRENT_LOG_LEVEL = level;
}

File: android/app/src/main/java/app/roam/android/util/Logger.kt
package app.roam.android.util

import android.util.Log
import io.sentry.Sentry
import app.roam.android.BuildConfig

/**
 * Centralized logging utility for Roam Android app.
 *
 * Prevents sensitive data leakage to:
 * 1. Logcat (visible via adb logcat, accessible via USB)
 * 2. App storage (SharedPreferences backup)
 * 3. Crash logs
 *
 * All errors automatically sent to Sentry for production monitoring.
 *
 * Usage:
 *   Logger.debug("auth", "Session loaded")
 *   Logger.error("roam", "Failed to fetch URL", statusCode = 404)
 */

object Logger {
  enum class Level(val priority: Int) {
    DEBUG(0),
    INFO(1),
    WARN(2),
    ERROR(3),
  }

  private var currentLevel = if (BuildConfig.DEBUG) Level.DEBUG else Level.ERROR

  /**
   * Debug level log (only in debug builds).
   */
  fun debug(tag: String, message: String, context: Map<String, Any>? = null) {
    if (currentLevel > Level.DEBUG) return
    val msg = formatMessage(tag, message, context)
    Log.d("Roam:$tag", msg)
  }

  /**
   * Info level log.
   */
  fun info(tag: String, message: String, context: Map<String, Any>? = null) {
    if (currentLevel > Level.INFO) return
    val msg = formatMessage(tag, message, context)
    Log.i("Roam:$tag", msg)
    Sentry.captureMessage("[INFO] [$tag] $message", io.sentry.SentryLevel.INFO)
  }

  /**
   * Warning level log.
   */
  fun warn(tag: String, message: String, context: Map<String, Any>? = null) {
    if (currentLevel > Level.WARN) return
    val msg = formatMessage(tag, message, context)
    Log.w("Roam:$tag", msg)
    Sentry.captureMessage("[WARN] [$tag] $message", io.sentry.SentryLevel.WARNING)
  }

  /**
   * Error level log (always captured to Sentry).
   */
  fun error(
    tag: String,
    message: String,
    context: Map<String, Any>? = null,
    throwable: Throwable? = null
  ) {
    val msg = formatMessage(tag, message, context)
    if (throwable != null) {
      Log.e("Roam:$tag", msg, throwable)
      Sentry.captureException(throwable) { scope ->
        scope.setTag("module", tag)
        scope.setTag("message", message)
        context?.forEach { (k, v) ->
          scope.setExtra(k, v.toString())
        }
      }
    } else {
      Log.e("Roam:$tag", msg)
      Sentry.captureMessage("[ERROR] [$tag] $message", io.sentry.SentryLevel.ERROR)
    }
  }

  /**
   * Override log level at runtime (for debugging).
   */
  fun setLevel(level: Level) {
    currentLevel = level
  }

  /**
   * Format a log message with sanitized context.
   * Safe fields: statusCode, count, duration, retry, attempt, reason
   * Unsafe fields (never logged): email, userId, password, token, url, response
   */
  private fun formatMessage(tag: String, message: String, context: Map<String, Any>?): String {
    if (context == null || context.isEmpty()) {
      return "[$tag] $message"
    }

    val safe = sanitizeContext(context)
    return if (safe.isEmpty()) {
      "[$tag] $message"
    } else {
      "[$tag] $message ${safe.entries.joinToString(", ") { "${it.key}=${it.value}" }}"
    }
  }

  private fun sanitizeContext(context: Map<String, Any>): Map<String, Any> {
    val SAFE_KEYS = setOf("statusCode", "count", "duration", "retry", "attempt", "reason", "category", "type", "action")
    val UNSAFE_KEYS = setOf("email", "password", "token", "secret", "userId", "id", "url", "response", "body", "payload", "session")

    return context.filter { (key, _) ->
      val keyLower = key.lowercase()
      UNSAFE_KEYS.none { keyLower.contains(it) } && SAFE_KEYS.any { keyLower.contains(it) }
    }
  }
}
